[samuelim01] iP

samuelim01 opened the pull request:
# Duke

> "A gamechanger." - said no one ever ([source](https://www.youtube.com/watch?v=dQw4w9WgXcQ))

Duke revolutionizes the todo experience. It's,
- text-based
- ~~easy to use~~ picky
- efficient

To use Duke, just
1. download the jar file
2. use the `java -jar duke.jar` command

Best part, it's **FREE**! 🤡 

Features:
- [X] Manage tasks
- [ ] GUI-based

As a throwback,
```java
public static <T> InfiniteList<T> generate(Producer<T> producer) {
  return new InfiniteList<>(
    Lazy.of(() -> Maybe.some(producer.produce())),
    Lazy.of(() -> InfiniteList.generate(producer)));
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

samuelim01 committed:
Implemented Level 0. Rename, Greet, Exit

samuelim01 committed:
Implemented Level 1. Echo

samuelim01 committed:
Implemented Level 2. Add, List

samuelim01 committed:
Implemented Level 3. Mark as Done

samuelim01 committed:
Implemented Level 4. ToDos, Events, Deadlines

samuelim01 committed:
Implemented A-TextUiTesting

samuelim01 committed:
Implemented Level 5. Handle Errors

samuelim01 committed:
Implemented Level 6. Delete

samuelim01 committed:
Implement Level-7: Save

samuelim01 committed:
Merge branch 'branch-Level-7'

samuelim01 committed:
Implement Level 8. Dates and Times

samuelim01 committed:
Merge branch 'branch-Level-8'

samuelim01 committed:
Implement Ui

samuelim01 committed:
Implement Storage

samuelim01 committed:
Implement TaskList

samuelim01 committed:
Implement Command and Parser

samuelim01 committed:
Merge branch 'branch-A-MoreOOP'

samuelim01 committed:
Implement A-Packages

samuelim01 committed:
Merge branch 'add-gradle-support'

samuelim01 committed:
Implement A-JUnit

samuelim01 committed:
Implemented A-JavaDoc

samuelim01 committed:
Implement Level-9

samuelim01 committed:
Implement A-CodingStandard

samuelim01 committed:
Merge branch 'branch-A-JavaDoc'

samuelim01 committed:
Merge branch 'branch-A-CodingStandard'

samuelim01 committed:
Merge branch 'branch-Level-9'

samuelim01 committed:
Implemented A-CheckStyle

samuelim01 committed:
Merge branch 'branch-A-CheckStyle'

garylow2001 commented on .gitignore:
```diff
@@ -13,5 +13,11 @@ src/main/resources/docs/
+*.class
+
+# Data
+/data/
```
Good job on git ignoring your data file, a common mistake many students make including me!

garylow2001 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,180 @@
+    private Command prepareMark(String arguments, boolean isMarked) throws DukeException {
+        final Matcher matcher = TASK_INDEX_ARGS_FORMAT.matcher(arguments.trim());
+        if (!matcher.matches()) {
+            throw new DukeException("The task must be specified.");
```
Perhaps you can abstract out 
```
DukeException("The task must be specified.")
```
to
```
InvalidTaskException
```
which extends from DukeException

garylow2001 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,109 @@
+                    task = new Event(description, localFrom, localTo);
+                    break;
+                default:
+                    throw new DukeException("Invalid save data.");
```
Similarly, you can create a ```InvalidSaveDataException``` that extends from ```DukeException```

garylow2001 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,109 @@
+     * @param tasks The list of tasks to be saved to the file.
+     * @throws DukeException If there was an issue saving the file.
+     */
+    public void save(TaskList tasks) throws DukeException {
```
```save``` is a very ambiguous method name. Perhaps you can change it to ```saveTask```

garylow2001 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,123 @@
+     *
+     * @param tasks The tasks to be printed.
+     */
+    public void showList(List<? extends Task> tasks) {
```
```showList``` can be quite ambiguous, perhaps ```showTaskList``` might be more specific

garylow2001 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,123 @@
+            System.out.println("\t Here are the matching tasks in your list:");
+        }
+
+        for (int i = 0; i < tasks.size(); i++) {
```
This is the same as ```showList```. Thus to prevent reusing code, perhaps you can replace these few lines of code after the else statement with
```
showList(tasks);
```

garylow2001 reviewed (commented):
LGTM. Overall, your code follows the coding standard closely, just some nits I have suggested!

jylow commented on src/main/java/duke/command/AddTodoCommand.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        tasks.add(toAdd);
+        ui.showMessage(
+                "Got it. I've added this task:",
+                "\t" + toAdd,
+                "Now you have " + tasks.size() + " tasks in the list."
+        );
+        storage.save(tasks);
+    }
+}
```
execute method is the same for at least 3 classes. Maybe you could try creating a default method in the ui to reduce duplication of code

jylow replied:
execute has code duplication in 3 different classes. Perhaps you could add a function to ui or even combine the 3 classes into a single add command

jylow commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,180 @@
+ */
+public class Parser {
+
+    private static final Pattern BASIC_COMMAND_FORMAT = Pattern.compile("(?<commandWord>\\S+)(?<arguments>.*)");
+    private static final Pattern TASK_INDEX_ARGS_FORMAT = Pattern.compile("(?<taskIndex>-?\\d+)");
+    private static final Pattern TODO_DATA_ARGS_FORMAT = Pattern.compile("(?<description>[^/]+)");
+    private static final Pattern DEADLINE_DATA_ARGS_FORMAT =
+            Pattern.compile("(?<description>[^/]+)"
+                    + " /by (?<by>[^/]+)");
+    private static final Pattern EVENT_DATA_ARGS_FORMAT =
+            Pattern.compile("(?<description>[^/]+)"
+                    + " /from (?<from>[^/]+)"
+                    + " /to (?<to>[^/]+)");
+    private static final Pattern KEYWORD_DATA_ARGS_FORMAT = Pattern.compile("(?<keyword>\\w+)");
```
Nice clean and concise way to save the different formats. I think I will try and use this in my code

jylow replied:
Nice clean and concise way to save the different formats. I think I will try and use this in my code

jylow commented on src/main/java/duke/command/AddEventCommand.java:
```diff
@@ -0,0 +1,53 @@
+        try {
+            this.toAdd = new Event(description, LocalDate.parse(from), LocalDate.parse(to));
+        } catch (DateTimeParseException e) {
+            throw new DukeException("The dates must be filled in \"yyyy-mm-dd\" format.");
```
Instead of throwing DukeException({Message}), it might be be more convenient to create a invalidDateFormat exception to reduce code duplication

jylow commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+                String fullCommand = ui.getUserCommand();
+                Command command = parser.parse(fullCommand);
+                command.execute(tasks, ui, storage);
+                isExit = command.isExit();
```
I was suggested by my reviewer to combine these 2 lines since execute returns void. Maybe it can just return if it is an exit command instead. I thought it made sense to me and would like to share with you

jylow reviewed (commented):
I think I learned quite a bit from reviewing your code. The way you structure your code might be of help to better organise mine. However, there is some code duplication that you could abstract out and make the code look cleaner

samuelim01 committed:
Implement Level-10

samuelim01 committed:
Improve use of OOP

samuelim01 committed:
Merge branch 'branch-Level-10'

samuelim01 committed:
Implement assertions

The current chatbot has no assertions. There may be instances where classes can be initialized without obeying some assumptions of the use of their constructors.

This could result in faulty code in the future due to the unchecked assumptions.

Let's implement assertions in duke.data.task.Task and duke.command.FindCommand.

Assertions will help to define, maintain, and verify important assumptions made by the codebase.

samuelim01 committed:
Delete Ui.java

The Ui class is no longer being used by Duke.

Removing the class will improve code quality as the codebase will only contain necessary code.

Let's delete Ui.java

samuelim01 committed:
Task modification commands: extract a parent class

Task modification command classes (DeleteCommand, MarkCommand, UnmarkCommand) have some common behaviour.

The common behaviour across classes result in code duplication.

Extracting the behaviour into a super class allows us to use polymorphism.

Let's create a new parent class ModifyTaskCommand.

Using inheritance will reduce the amount of duplicated code.

samuelim01 committed:
Add StorageSaveException

samuelim01 committed:
Made task and command fields final

samuelim01 committed:
Merge pull request #3 from samuelim01/branch-A-Assertions

Implement assertions

samuelim01 committed:
Merge branch 'master' into branch-A-CodeQuality

samuelim01 committed:
Merge pull request #4 from samuelim01/branch-A-CodeQuality

Improve code quality

samuelim01 committed:
Create gradle.yml

samuelim01 committed:
Merge pull request #5 from samuelim01/branch-A-CI

Implement A-CI

samuelim01 committed:
Implement HelpCommand

samuelim01 committed:
Abstract command descriptions as class variables

samuelim01 committed:
Merge branch 'branch-C-Help'

samuelim01 committed:
Add CommandResult abstraction

samuelim01 committed:
Add additional Parser testcases

samuelim01 committed:
Add user guide

samuelim01 committed:
build.gradle: Add dependencies

The iP JAR file crashes on Linux

Let's follow https://github.com/nus-cs2103-AY2324S1/forum/issues/173

samuelim01 committed:
Refactor `Duke` to `Socrates`
